src/main/java/org/dependencytrack/resources/v1/BomResource.java

```diff
@@ -454,7 +455,26 @@ public Response uploadBom(
                         return Response.status(Response.Status.UNAUTHORIZED).entity("The principal does not have permission to create project.").build();
                     }
                 }
-                return process(qm, project, artifactParts);
+                final var response = process(qm, project, artifactParts);
+                if (response.getStatusInfo() == Response.Status.OK && isActiveExclusively && trimmedProjectName != null) {
+                    LOGGER.info("Deactivating old versions for project: " + trimmedProjectName);
+                    if (!isLatest) {
+                        var message = "Value \"isLatest=true\" required when \"isActiveExclusively=true\".";
+                        LOGGER.error(message);
+                        return Response.status(Response.Status.NOT_ACCEPTABLE).entity(message).build();
+                    }
+                    qm.getProjects(trimmedProjectName, true, false, null).getList(Project.class).forEach(p -> {
+                        if (p.isLatest()) {
+                            return;
+                        } else if (!qm.hasAccess(super.getPrincipal(), p)) {
+                            LOGGER.warn("Could not deactivate project, no access: " + p.getUuid() + " / " + p.getName());
+                            return;
+                        }
+                        p.setActive(false);
+                        qm.updateProject(p, true);
+                    });
+                }
+                return response;
```

Comment:
Should this be atomic with the new project being created (see lines `451-453`), and happen *before* them BOM is processed?

At the moment, deactivating other projects could fail, resulting in a `HTTP 500` response, despite the new project having been created successfully, and the BOM processing having been triggered successfully.

You can use `qm.runInTransaction(() -> { /* your code */ })` to execute a portion of code in a transaction. You can find examples of this pattern throughout the code base.

Reply:
Good call, made this change a lot prettier as well.